File: 130-surrounded-regions/surrounded-regions.py
from typing import List
class Solution:
    def __init__(self):
        self.ROWS = 0
        self.COLS = 0
        self.directions = []
    def isValid(self, r, c): return 0 <= r < self.ROWS and 0 <= c < self.COLS
    def solve(self, board: List[List[str]]) -> None:
        self.directions = [(1,0),(-1,0),(0,1),(0,-1)]
        self.ROWS = len(board)
        self.COLS = len(board[0])
        def dfs(R,C):
            if not self.isValid(R,C): return
            if board[R][C] in {'T','X'}: return
            board[R][C] = 'T'
            for x,y in self.directions:
                dfs(R+x,C+y)
        for i in range(self.ROWS):
            if board[i][0] == 'O': dfs(i,0)
            if board[i][self.COLS-1] == 'O': dfs(i,self.COLS-1)
        for i in range(self.COLS):
            if board[0][i] == 'O': dfs(0,i)
            if board[self.ROWS-1][i] == 'O':dfs(self.ROWS-1,i)
        for row in range(self.ROWS):
            for col in range(self.COLS):
                if board[row][col] == 'O': board[row][col] = 'X'
                if board[row][col] == 'T': board[row][col] = 'O'

File: 130-surrounded-regions/surrounded-regions.cpp
class Solution {
public:
    vector<vector<int>> dir = { {1,0}, {-1,0}, {0,1}, {0,-1}};
    void dfs(int row, int col, vector<vector<char>>& grid,vector<vector<bool>> &visited){
        visited[row][col] = true;
        for (int i =0; i<4; i++){
            int x = row + dir[i][0];
            int y = col + dir[i][1];
            if (x>=0 and x<grid.size() and y>=0 and y<grid[0].size() and grid[x][y] == 'O' and !visited[x][y]){ dfs(x,y,grid,visited); }
        }
    }
    void solve(vector<vector<char>>& grid) {
        int ROWS= grid.size();
        int COLS= grid[0].size();
        vector<vector<bool>> visited (ROWS,vector<bool>(COLS,false));


        for (int col = 0; col<COLS; col++){
           if (grid[0][col] == 'O'){ dfs(0,col,grid,visited); }
        }
        for (int col = 0; col<COLS; col++){
           if (grid[ROWS-1][col] == 'O'){ dfs(ROWS-1,col,grid,visited); }
        }
        for (int row = 0; row<ROWS; row++){
           if (grid[row][0] == 'O'){ dfs(row,0,grid,visited); }
        }
        for (int row = 0; row<ROWS; row++){
           if (grid[row][COLS-1] == 'O'){ dfs(row,COLS-1,grid,visited); }
        }


        for (int row = 1; row<ROWS-1; row++){
            for (int col = 1; col<COLS-1; col++){
                if(grid[row][col] == 'O' && !visited[row][col]) grid[row][col] = 'X';
            }
        }
    }
};